File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Foundation/ErrorTypes/ErrorContext.ts
/**
 * ErrorContext.ts
 *
 * Captures detailed error context for debugging and logging.
 * Matches iOS SDK: Foundation/ErrorTypes/ErrorContext.swift
 */

/**
 * Contextual information captured when an error occurs.
 */
export interface ErrorContext {
  /** Stack trace at error point */
  readonly stackTrace: string[];
  /** Source file where error occurred */
  readonly file: string;
  /** Line number */
  readonly line: number;
  /** Function name */
  readonly function: string;
  /** Error capture timestamp (ISO8601) */
  readonly timestamp: string;
  /** Thread info ("main" or "background") */
  readonly threadInfo: string;
}

/**
 * Create an error context from the current call site.
 * Note: In JavaScript, we can only capture stack traces, not file/line/function directly.
 */
export function createErrorContext(error?: Error): ErrorContext {
  const now = new Date().toISOString();
  const stackTrace = parseStackTrace(error?.stack ?? new Error().stack ?? '');

  // Extract location from first relevant stack frame
  const location = extractLocationFromStack(stackTrace);

  return {
    stackTrace,
    file: location.file,
    line: location.line,
    function: location.function,
    timestamp: now,
    threadInfo: 'main', // JS is single-threaded (main thread)
  };
}

/**
 * Parse a stack trace string into an array of frames.
 */
function parseStackTrace(stack: string): string[] {
  const lines = stack.split('\n');

  return lines
    .slice(1) // Skip "Error: message" line
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .filter((line) => !isSystemFrame(line))
    .slice(0, 15); // Limit to 15 frames like iOS
}

/**
 * Check if a stack frame is a system/internal frame that should be filtered.
 */
function isSystemFrame(frame: string): boolean {
  const systemPatterns = [
    'node_modules',
    'internal/',
    '__webpack',
    'regenerator',
    'asyncToGenerator',
    'createErrorContext', // Filter ourselves out
    'parseStackTrace',
  ];

  return systemPatterns.some((pattern) => frame.includes(pattern));
}

/**
 * Extract file, line, and function from the first relevant stack frame.
 */
function extractLocationFromStack(stackTrace: string[]): {
  file: string;
  line: number;
  function: string;
} {
  if (stackTrace.length === 0) {
    return { file: 'unknown', line: 0, function: 'unknown' };
  }

  const firstFrame = stackTrace[0];

  // Try to parse "at functionName (file:line:column)" format
  const atMatch = firstFrame.match(/at\s+(.+?)\s+\((.+?):(\d+):(\d+)\)/);
  if (atMatch) {
    return {
      function: atMatch[1],
      file: atMatch[2],
      line: parseInt(atMatch[3], 10),
    };
  }

  // Try to parse "at file:line:column" format (anonymous function)
  const atFileMatch = firstFrame.match(/at\s+(.+?):(\d+):(\d+)/);
  if (atFileMatch) {
    return {
      function: 'anonymous',
      file: atFileMatch[1],
      line: parseInt(atFileMatch[2], 10),
    };
  }

  // Try to parse "functionName@file:line:column" format (Safari/Firefox)
  const atSignMatch = firstFrame.match(/(.+?)@(.+?):(\d+):(\d+)/);
  if (atSignMatch) {
    return {
      function: atSignMatch[1] || 'anonymous',
      file: atSignMatch[2],
      line: parseInt(atSignMatch[3], 10),
    };
  }

  return { file: 'unknown', line: 0, function: 'unknown' };
}

/**
 * Format the stack trace as a readable string.
 */
export function formatStackTrace(context: ErrorContext): string {
  if (context.stackTrace.length === 0) {
    return 'No stack trace available';
  }
  return context.stackTrace.join('\n');
}

/**
 * Get a formatted location string (file:line in function).
 */
export function formatLocation(context: ErrorContext): string {
  return `${context.file}:${context.line} in ${context.function}`;
}

/**
 * Get a complete formatted context string for logging.
 */
export function formatContext(context: ErrorContext): string {
  return [
    `Time: ${context.timestamp}`,
    `Thread: ${context.threadInfo}`,
    `Location: ${formatLocation(context)}`,
    `Stack Trace:`,
    formatStackTrace(context),
  ].join('\n');
}

/**
 * An error wrapper that includes context information.
 */
export class ContextualError extends Error {
  readonly context: ErrorContext;
  readonly originalError: Error;

  constructor(error: Error, context?: ErrorContext) {
    super(error.message);
    this.name = 'ContextualError';
    this.originalError = error;
    this.context = context ?? createErrorContext(error);

    // Maintain proper prototype chain
    Object.setPrototypeOf(this, ContextualError.prototype);
  }
}

/**
 * Wrap an error with context information.
 */
export function withContext(error: Error): ContextualError {
  if (error instanceof ContextualError) {
    return error; // Already has context
  }
  return new ContextualError(error);
}

/**
 * Extract error context from an error if available.
 */
export function getErrorContext(error: Error): ErrorContext | undefined {
  if (error instanceof ContextualError) {
    return error.context;
  }
  return undefined;
}

/**
 * Get the underlying error value, unwrapping ContextualError if needed.
 */
export function getUnderlyingError(error: Error): Error {
  if (error instanceof ContextualError) {
    return error.originalError;
  }
  return error;
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/foundation/error_types/error_context.dart
/// Context information captured when an error occurs
/// Includes stack trace, source location, and timing information
/// Matches iOS ErrorContext from Foundation/ErrorTypes/ErrorContext.swift
class ErrorContext {
  /// The stack trace at the point of error capture
  final StackTrace? stackTrace;

  /// The file where the error was captured
  final String file;

  /// The line number where the error was captured
  final int line;

  /// The function where the error was captured
  final String function;

  /// Timestamp when the error was captured
  final DateTime timestamp;

  /// Thread name/ID where the error occurred
  final String threadInfo;

  /// Initialize with automatic capture of current context
  ErrorContext({
    this.stackTrace,
    this.file = '',
    this.line = 0,
    this.function = '',
    DateTime? timestamp,
    String? threadInfo,
  })  : timestamp = timestamp ?? DateTime.now(),
        threadInfo = threadInfo ?? 'main';

  /// Create context from stack trace
  factory ErrorContext.capture() {
    final trace = StackTrace.current;
    return ErrorContext(
      stackTrace: trace,
      timestamp: DateTime.now(),
      threadInfo: 'main',
    );
  }

  /// Initialize with explicit values (for testing or deserialization)
  factory ErrorContext.explicit({
    required StackTrace? stackTrace,
    required String file,
    required int line,
    required String function,
    required DateTime timestamp,
    required String threadInfo,
  }) {
    return ErrorContext(
      stackTrace: stackTrace,
      file: file,
      line: line,
      function: function,
      timestamp: timestamp,
      threadInfo: threadInfo,
    );
  }

  /// A formatted string representation of the stack trace
  String get formattedStackTrace {
    if (stackTrace == null) return '';

    final lines = stackTrace.toString().split('\n');
    final relevantFrames = lines
        .where(
            (frame) => frame.contains('runanywhere') || frame.contains('lib/'))
        .take(15)
        .toList();

    if (relevantFrames.isEmpty) {
      return lines.take(10).join('\n');
    }

    return relevantFrames
        .asMap()
        .entries
        .map((e) => '  ${e.key}. ${e.value}')
        .join('\n');
  }

  /// A compact single-line location string
  String get locationString => '$file:$line in $function';

  /// Full formatted context for logging
  String get formattedContext => '''
Location: $locationString
Thread: $threadInfo
Time: ${timestamp.toIso8601String()}
Stack Trace:
$formattedStackTrace
''';

  /// Convert to map for serialization
  Map<String, dynamic> toJson() => {
        'file': file,
        'line': line,
        'function': function,
        'timestamp': timestamp.toIso8601String(),
        'threadInfo': threadInfo,
        'stackTrace': stackTrace?.toString(),
      };

  /// Create from map
  factory ErrorContext.fromJson(Map<String, dynamic> json) {
    return ErrorContext(
      file: json['file'] as String? ?? '',
      line: json['line'] as int? ?? 0,
      function: json['function'] as String? ?? '',
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'] as String)
          : DateTime.now(),
      threadInfo: json['threadInfo'] as String? ?? 'main',
    );
  }
}

/// Global function to capture error context at the call site
/// Use this when throwing errors to capture the stack trace
ErrorContext captureErrorContext() => ErrorContext.capture();

/// A wrapper that attaches context to any error
class ContextualError implements Exception {
  /// The underlying error
  final Object error;

  /// The captured context
  final ErrorContext context;

  /// Initialize with an error and automatically capture context
  ContextualError(this.error, {ErrorContext? context})
      : context = context ?? ErrorContext.capture();

  @override
  String toString() {
    final errorDesc =
        error is Exception ? (error as Exception).toString() : error.toString();
    return errorDesc;
  }

  /// Get error description
  String? get errorDescription {
    if (error is Exception) {
      return error.toString();
    }
    return error.toString();
  }
}

/// Extension to add context to any error
extension ErrorContextExtension on Object {
  /// Wrap this error with context information
  ContextualError withContext() => ContextualError(this);

  /// Extract context if this is a ContextualError
  ErrorContext? get errorContext {
    if (this is ContextualError) {
      return (this as ContextualError).context;
    }
    return null;
  }

  /// Get the underlying error if wrapped
  Object get underlyingErrorValue {
    if (this is ContextualError) {
      return (this as ContextualError).error;
    }
    return this;
  }
}
